Updates Fungible Token tutorial to use a different structure and explainations (#1238)
onflow/cadence: docs/tutorial/03-fungible-tokens.mdx

Some of the most popular contract classes on blockchains today are fungible tokens.
These contracts create homogeneous tokens that can be transferred to other users and spent as currency (e.g., ERC-20 on Ethereum).

In traditional software and smart contracts, balances for each user are tracked by a central ledger, such as a dictionary:
```cadence
contract LedgerToken {
    // Tracks every user's balance

>     pub let balances: {Address: UFix64}

Reviewer: I know this is "don't do this code" but this dictionary is publicly mutatable. Maybe  make it`access(contract) let` ?

Author: Good point. I'll change that

</Callout>

> The concepts involved in implementing a fungible token in Cadence can be hard to grasp at first.

Reviewer: ```suggestion
The concepts involved in implementing a fungible token in Cadence can be unfamiliar at first.
```

pub resource Vault: Provider, Receiver {

    // Balance of a user's Vault
    // we use unsigned fixed point numbers for balances 

>     // because they can represent decimals and do not require the concept of a negative number

Reviewer: ```suggestion
    // because they can represent decimals and do not allow negative values
```

}
```

### Transferring Tokens

When an account wants to send tokens to a different account, the sending account calls their own withdraw function first,
which subtracts tokens from their resource’s balance and temporarily creates a new resource object that holds this balance.
The sending account then calls the recipient account’s deposit function, which literally moves the resource instance to the other account,
adds it to their balance, and then destroys the used resource.
The resource needs to be destroyed because Cadence enforces strict rules around resource interactions.
A resource can never be left hanging in a piece of code. It either needs to be explicitly destroyed or stored in an account's storage.

> When interacting with resources, you use the `@` symbol to specify the type and a special “move operator” `<-` in assignments.

Reviewer: 👍

```

In Solidity, this could be a risk for integer overflow or underflow, but Cadence has built-in overflow and underflow protection, so it is not a risk.

> We are also using unsigned numbers in this example, so the vault`s balance cannot go below 0.

Reviewer: We mention this advantage above, so if we wish to re-mention it we could say "As we mentioned above...".

        ?? panic("Could not borrow a reference to the owner's vault")

    let temporaryVault <- vaultRef.withdraw(amount: 10.0)

    // deposit your tokens to the Vault
    vaultRef.deposit(from: <-temporaryVault)

    log("Withdraw/Deposit succeeded!")
  }
}
```

<Callout type="info">
Select account `0x01` as the only signer. <br/>
Click the `Send` button to submit the transaction. <br/>
This transaction withdraws tokens from the main vault and deposits them back to it.
</Callout>

This transaction is a basic example of a transfer within an account.
It withdraws tokens from the main vault and deposits back to the main vault.
It is simply to illustrate the basic functionality of how transfers work.

You'll see in this transaction that
you can borrow a reference directly from an object in storage.

```cadence
// Borrow a reference to the stored, private Vault resource
let vaultRef = acct.borrow<&BasicToken.Vault>(from: /storage/CadenceFungibleTokenTutorialVault)
    ?? panic("Could not borrow a reference to the owner's vault")
```

This allows you to efficiently access objects in storage without having to load them,
which is a much more costly interaction.

> In the real world, you'll likely be transferring tokens to other accounts.

Reviewer: ```suggestion
In production code, you'll likely be transferring tokens to other accounts.
```

Capabilities allow us to accomplish this safely.

## Ensuring Security in Public: Capability Security

---

Another important feature in Cadence is its utilization of **Capability Security.**

> This feature ensures that, while the withdraw function is public,

Reviewer: ```suggestion
This feature ensures that, while the withdraw function is declared public on the resource,
```

no one except the intended user and those they approve of can withdraw tokens from their vault.

Cadence's security model ensures that objects stored in an account's storage can only be accessed by the account that owns them.
If a user wants to give another user access to their stored objects, they can link a public capability,
which is like an "API" that allows others to call specified functions on their objects.

An account only has access to the fields and methods of an object in a different account if they hold a capability to that object
that explicitly allows them to access those fields and methods. 
Only the owner of an object can create a capability for it.
Therefore, when a user creates a Vault in their account, they only publish a capability
that exposes the deposit function and the balance field.

The withdraw function can remain hidden as a function that only the owner can call.

> As you can hopefully see, this removes the need to check `msg.sender` for access control purposes,

Reviewer: ```suggestion
This removes the need to check `msg.sender` for access control purposes,
```

Reviewer 2: nit: This seems to be the first time that 'msg.sender' appears in this series of tutorial. Perhaps worth to mention this is the Ethereum idiom to handle access control

Author: Good point! I made that change

Here is an example of how interfaces for the `Vault` resource we defined above would look.

```cadence:title=Interfaces.cdc
// Interface that enforces the requirements for withdrawing
// tokens from the implementing type
//
pub resource interface Provider {
    pub fun withdraw(amount: UFix64): @Vault {
        post {
            result.balance == amount:
                "Withdrawal amount must be the same as the balance of the withdrawn Vault"
        }
    }
}
// Interface that enforces the requirements for depositing
// tokens into the implementing type
//
pub resource interface Receiver {

    // There aren't any meaningful requirements for only a deposit function
    // but this still shows that the deposit function is required in an implementation.
    pub fun deposit(from: @Vault)
}

// Balance
//
// Interface that specifies a public `balance` field for the vault
//
pub resource interface Balance {
    pub var balance: UFix64
}
```

In our example, the `Vault` resource will implement all three of these interfaces.
The interfaces ensure that specific fields and functions are present in the resource implementation

> and that those functions meet certain conditions before and/or after execution.

Reviewer: ```suggestion
and that the function arguments, fields of the resource, and any return value are in a valid state before and/or after execution.
```

These interfaces can be stored on-chain and imported into other contracts or resources
so that these requirements are enforced by an immutable source of truth that is not susceptible to human error.

You can also see that functions and fields have the `pub` keyword next to them.
We have explicitly defined these fields as public because all fields and functions in Cadence are private by default,
meaning that the local scope can only access them. Users have to make parts of their owned types explicitly public.
This helps prevent types from having unintentionally public code.

This does bring up an important security consideration though.
While we have made all our fields and functions public here, it is actually recommended to default to making
fields private unless it is explicitly needed to be public.

> This is especially important for array and dictionary types, which can have their fields maliciously assigned to

Reviewer: ```suggestion
This is especially important for array and dictionary types, which can have their contents maliciously mutated
```

be evaluated every time their respective functions are called.
We can also use these interfaces to create restricted capabilities, described in the next section.

Additionally, `ExampleToken` changes `createVault()` to `createEmptyVault()`
so that token minting is restricted to the newly added `VaultMinter` resource.
This illustrates another powerful feature of Cadence resources.
Instead of the contract maintaining a list of minter addresses,
accounts that the owner wants to be minters can be giving a special resource
that directly gives them the authority to mint tokens.
This method for authorization can be used in many different ways
and further decentralizes the control of the contract.

We also store the `VaultMinter` object to `/storage/`
in the `init()` function in the same way as the vault, but in a different storage path:

```cadence
self.account.save(<-create VaultMinter(), to: /storage/CadenceFungibleTokenTutorialMinter)
```

Now is an important time to remind you that account storage not namespaced by contract,
meaning that path names could potentially conflict. This is why it is important

> choose unique names for your paths like we have done here so there is a very low chance

Reviewer: ```suggestion
to choose unique names for your paths like we have done here so there is a very low chance
```

}
```

The function takes a capability as an argument. 

> This syntax might be confusing to you:

Reviewer: ```suggestion
This syntax might be unclear to you:
```
